### PR TITLE
Improve elementwise operators performance in same dimensions.

### PR DESCRIPTION
**Pull request by danleifeng**
changes are included:
1. add mkl(for cpu) and cuda kernel(for gpu) codes  for performance in same dimensions.(add/sub/div)
2. import cuda math api(__hadd2, etc)  for float16 performance in same dimensions.(add/sub/mul/div)
3. use macro define and template functions for code conciseness.

### CODE REVIEW

**Comment by luotao1** [comment]
It's better to separate this PR into two PRs, one for CPU, and one for GPU.

**Comment by zhaoyuchen2018** [comment]
Could you please give some performance data?  
@kuke   maybe  can check performance improves on your model.

File: paddle/fluid/operators/elementwise/elementwise_add_op.cc [hunk 1]
```diff
@@ -20,6 +20,41 @@ limitations under the License. */
...
+};
+
+template <typename T>
+struct SameDimsElemwiseAdd<
+    platform::CPUDeviceContext, T,
+    typename std::enable_if<!std::is_floating_point<T>::value>::type> {
+  void operator()(const framework::ExecutionContext &ctx,
+                  const framework::Tensor *x, const framework::Tensor *y,
+                  framework::Tensor *z) {
+    VLOG(5) << "====into eigen ===";
```
Comment (gongweibao) [inline paddle/fluid/operators/elementwise/elementwise_add_op.cc]: Delete these vlog after debug!

File: paddle/fluid/operators/elementwise/elementwise_op_function.cu.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+#pragma once
+
+#include <glog/logging.h>
+#include "paddle/fluid/operators/elementwise/elementwise.h"
+
+namespace paddle {
+namespace operators {
+
+#if defined(__CUDACC__) && CUDA_VERSION >= 7050
+#define PADDLE_CUDA_FP16
```
Comment (gongweibao) [inline paddle/fluid/operators/elementwise/elementwise_op_function.cu.h]: duplicate code with float16.h.

Comment (gongweibao) [reply paddle/fluid/operators/elementwise/elementwise_op_function.cu.h]: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/platform/float16.h#L36

File: paddle/fluid/operators/elementwise/elementwise_mul_op.cu [hunk 3]
```diff
@@ -1,18 +1,16 @@
...
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License. */
 
 #include "paddle/fluid/operators/elementwise/elementwise_mul_op.h"
+#include "paddle/fluid/operators/elementwise/elementwise_op_function.cu.h"
 #include "paddle/fluid/platform/float16.h"
 
 #define TILE_SIZE 512
```
Comment (gongweibao) [inline paddle/fluid/operators/elementwise/elementwise_mul_op.cu]: Duplicate defination.

File: paddle/fluid/operators/elementwise/elementwise_add_op.cu [hunk 4]
```diff
@@ -1,23 +1,114 @@
...
+          dout->data<T>(), size, dx->mutable_data<T>(ctx.GetPlace()),
+          dy->mutable_data<T>(ctx.GetPlace()));
+    } else {
+      default_elementwise_add_grad<plat::CUDADeviceContext, T>(ctx, x, y, out,
+                                                               dout, dx, dy);
+    }
+  }
+};
+
+template struct SameDimsElemwiseAdd<platform::CUDADeviceContext, float>;
```
Comment (gongweibao) [inline paddle/fluid/operators/elementwise/elementwise_add_op.cu]: 下边这几行可以用宏，因为不是需要对用户可读友好的代码！

File: paddle/fluid/operators/elementwise/elementwise_div_op.cu [hunk 5]
```diff
@@ -1,20 +1,112 @@
...
+  while (col < size) {
+    T o = dout[col];
+    dx[col] = o / y[col];
+    dy[col] = -o * out[col] / y[col];
+    col += blockDim.x * gridDim.x;
+  }
+}
+
+template <typename T>
+class ElementwiseDivGradKernel<plat::CUDADeviceContext, T>
```
Comment (gongweibao) [inline paddle/fluid/operators/elementwise/elementwise_div_op.cu]: Clear this class!

File: paddle/fluid/operators/elementwise/elementwise_add_op.cu [hunk 6]
```diff
@@ -1,23 +1,111 @@
...
 
+namespace paddle {
+namespace operators {
+
+template <typename T>
+struct SameDimsElemwiseAdd<platform::CUDADeviceContext, T> {
+  void operator()(const framework::ExecutionContext& ctx,
+                  const framework::Tensor* x, const framework::Tensor* y,
+                  framework::Tensor* z) {
+    getAddFunctor<T> functor(x->data<T>(), y->data<T>(), z->data<T>());
```
Comment (gongweibao) [inline paddle/fluid/operators/elementwise/elementwise_add_op.cu]: AddFunctor?

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/elementwise/elementwise_add_op.cc (hunk 1 above)
Comment (danleifeng) [reply paddle/fluid/operators/elementwise/elementwise_add_op.cc]: done.

File: paddle/fluid/operators/elementwise/elementwise_mul_op.cu (hunk 3 above)
Comment (danleifeng) [reply paddle/fluid/operators/elementwise/elementwise_mul_op.cu]: done.

File: paddle/fluid/operators/elementwise/elementwise_add_op.cu (hunk 4 above)
Comment (danleifeng) [reply paddle/fluid/operators/elementwise/elementwise_add_op.cu]: done.

File: paddle/fluid/operators/elementwise/elementwise_div_op.cu (hunk 5 above)
Comment (danleifeng) [reply paddle/fluid/operators/elementwise/elementwise_div_op.cu]: done.

File: paddle/fluid/operators/math/blas_impl.h [hunk 7]
```diff
@@ -457,6 +491,22 @@ void Blas<platform::CPUDeviceContext>::VMUL(int n, const T *x, const T *y,
 #endif
 }
 
+template <>
+template <typename T>
+void Blas<platform::CPUDeviceContext>::VDIV(int n, const T *x, const T *y,
+                                            T *z) const {
+#ifdef PADDLE_WITH_MKLML
+  // VLOG(5) << "====into PADDLE_WITH_MKLML =>CBlas<T>::VDIV ===";
```
Comment (gongweibao) [inline paddle/fluid/operators/math/blas_impl.h]: Clear this and other comments!

File: paddle/fluid/operators/elementwise/elementwise_add_op.cu (hunk 6 above)
Comment (danleifeng) [reply paddle/fluid/operators/elementwise/elementwise_add_op.cu]: AddFunctor is defined in elementwise_op_function.cu.h
maybe AddRangeFunctor better.

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/elementwise/elementwise_op_function.h [hunk 8]
```diff
@@ -30,6 +30,7 @@ limitations under the License. */
 #include "paddle/fluid/platform/cuda_primitives.h"
 constexpr int ELEMWISE_MAX_BLOCK_DIM = 1024;
 #endif
+#define TILE_SIZE 512
```
Comment (gongweibao) [inline paddle/fluid/operators/elementwise/elementwise_op_function.h]: Is this needed?

File: paddle/fluid/operators/elementwise/elementwise_op_function.cu.h (hunk 2 above)
Comment (danleifeng) [reply paddle/fluid/operators/elementwise/elementwise_op_function.cu.h]: done.

File: paddle/fluid/operators/math/blas_impl.h (hunk 7 above)
Comment (danleifeng) [reply paddle/fluid/operators/math/blas_impl.h]: done.

File: paddle/fluid/operators/elementwise/elementwise_op_function.h (hunk 8 above)
Comment (danleifeng) [reply paddle/fluid/operators/elementwise/elementwise_op_function.h]: no need. 
already deleted.

**Review by gongweibao — APPROVED** [review]
LGTM

**Pull request merged**